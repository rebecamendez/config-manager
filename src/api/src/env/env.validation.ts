import { Expose, plainToInstance } from 'class-transformer';
import { IsBoolean, IsNumber, IsString, validateSync } from 'class-validator';

export class EnvironmentVariables {
  @Expose()
  @IsNumber()
  public readonly PORT: number;

  @Expose()
  @IsString()
  public readonly DB_HOST: string;

  @Expose()
  @IsNumber()
  public readonly DB_PORT: number;

  @Expose()
  @IsString()
  public readonly DB_USER: string;

  @Expose()
  @IsString()
  public readonly DB_PASSWORD: string;

  @Expose()
  @IsString()
  public readonly DB_DATABASE: string;

  @Expose()
  @IsBoolean()
  public readonly DB_ENABLE_LOGGER: boolean;
}

export const validate = (config: Record<string, unknown>): EnvironmentVariables => {
  const validatedConfig = plainToInstance(EnvironmentVariables, config, {
    enableImplicitConversion: true,
    excludeExtraneousValues: true
  });
  const errors = validateSync(validatedConfig, { skipMissingProperties: false });

  if (errors.length > 0) throw new Error(errors.toString());
  return validatedConfig;
};
