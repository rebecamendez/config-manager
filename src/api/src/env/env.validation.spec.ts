import 'reflect-metadata';
import { validate } from 'env/env.validation';

describe('An environment variables validation', () => {
  const mockEnv = {
    PORT: 3000,
    DB_HOST: 'localhost',
    DB_PORT: 5432,
    DB_USER: 'root',
    DB_PASSWORD: 'password',
    DB_DATABASE: 'config_manager_db',
    DB_ENABLE_LOGGER: true
  };

  it('should return env variables', async () => {
    const env = validate(mockEnv);
    expect(env).toMatchSnapshot();
  });

  it('should throw an error if has validation issues', async () => {
    expect(() => validate({})).toThrow(Error);
  });
});
