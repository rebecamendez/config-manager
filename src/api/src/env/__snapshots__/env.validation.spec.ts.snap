// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An environment variables validation should return env variables 1`] = `
EnvironmentVariables {
  "DB_DATABASE": "config_manager_db",
  "DB_ENABLE_LOGGER": true,
  "DB_HOST": "localhost",
  "DB_PASSWORD": "password",
  "DB_PORT": 5432,
  "DB_USER": "root",
  "PORT": 3000,
}
`;
